src/lib/provable/gadgets/blake2b.ts

```diff
@@ -0,0 +1,282 @@
+ * @param {State} state
+ * @param {boolean} last
+ */
+function compress(
+  state: {
+    h: UInt64[];
+    t: UInt64[];
+    buf: UInt8[];
+    buflen: number;
+    outlen: number;
+  },
+  last: boolean
+): State {
+  const { h, t, buf } = state;
+  const v = h.concat(BLAKE2B.IV); // initalize local work vector. First half from state and second half from IV.
+
+  v[12] = v[12].xor(t[0]); // low word of the offset
+  v[13] = v[13].xor(t[1]); // high word of the offset
+
+  if (last) {
+    // last block flag set ?
+    v[14] = v[14].not();
+  }
+
+  const m: UInt64[] = [];
+  for (let i = 0; i < 16; i++) {
+    // get little-endian words
+    m.push(
+      UInt64.fromFields(
+        buf[i * 8].value
+          .add(buf[i * 8 + 1].value.mul(1n << 8n))
+          .add(buf[i * 8 + 2].value.mul(1n << 16n))
+          .add(buf[i * 8 + 3].value.mul(1n << 24n))
+          .add(buf[i * 8 + 4].value.mul(1n << 32n))
+          .add(buf[i * 8 + 5].value.mul(1n << 40n))
+          .add(buf[i * 8 + 6].value.mul(1n << 48n))
+          .add(buf[i * 8 + 7].value.mul(1n << 56n)).toFields()
+      )
+    );
+  }
+
+  for (let i = 0; i < 12; i++) {
+    // twelve rounds
+    const s = BLAKE2BConstants.SIGMA[i % 10];
+    G(v, 0, 4, 8, 12, m[s[0]], m[s[1]]);
+    G(v, 1, 5, 9, 13, m[s[2]], m[s[3]]);
+    G(v, 2, 6, 10, 14, m[s[4]], m[s[5]]);
+    G(v, 3, 7, 11, 15, m[s[6]], m[s[7]]);
+
+    G(v, 0, 5, 10, 15, m[s[8]], m[s[9]]);
+    G(v, 1, 6, 11, 12, m[s[10]], m[s[11]]);
+    G(v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+    G(v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+  }
+
+  for (let i = 0; i < 8; i++) {
+    // XOR the two halves
+    h[i] = v[i].xor(v[i + 8]).xor(h[i]);
+  }
+  return state;
+}
+
+/**
+ * Initializes the state with the given digest length.
+ *
+ * @param {number} outlen - Digest length in bits
+ * @returns {State}
+ */
+function initialize(outlen: number): {
+  h: UInt64[];
+  t: UInt64[];
+  buf: UInt8[];
+  buflen: number;
+  outlen: number;
+} {
+  const h = BLAKE2B.IV.slice(); // shallow copy IV to h
+  h[0] = UInt64.from(0x01010000).xor(UInt64.from(outlen)).xor(h[0]); // state "param block"
+
+  return {
+    h,
+    t: [UInt64.zero, UInt64.zero],
+    buf: [],
+    buflen: 0,
+    outlen,
+  };
+}
+
+/**
+ * Updates hash state
+ * @param {State} state
+ * @param {UInt8[]} input
+ * @returns {State} updated state
+ */
+function update(
+  state: {
+    h: UInt64[];
+    t: UInt64[];
+    buf: UInt8[];
+    buflen: number;
+    outlen: number;
+  },
+  input: UInt8[]
+): State {
+  for (let i = 0; i < input.length; i++) {
+    if (state.buflen === 128) {
+      // buffer full ?
+      state.t[0] = state.t[0].addMod64(UInt64.from(state.buflen)); // add counters
+      if (state.t[0].toBigInt() < state.buflen) {
+        // carry overflow ?
+        state.t[1] = state.t[1].addMod64(UInt64.one); // high word
+      }
+      state = compress(state, false); // compress (not last)
+      state.buflen = 0; // counter to zero
```

Comment:
Shouldn't this also reset `state.buf` to an empty array? If you don't do this, you'll start overwriting `buf` at the start but the rest of it will still contain the old entries.

and if yes, wouldn't it be less confusing in general to always use `state.buf.length` instead of an extra parameter `buflen`, so that they can't get out of sync?

Reply:
The reference implementation doesn’t do that, so I followed the same approach.

Reply:
Ok interesting, in that case I would try to find a spec and see if it matches the reference impl on this point. It could indeed be the correct behavior, or it could be a bug in the reference!

Reply:
There are two specifications on the [official BLAKE2B website](https://www.blake2.net). [One of them](https://www.blake2.net/blake2.pdf) doesn’t include a reference implementation, while the [other](https://datatracker.ietf.org/doc/html/rfc7693#section-2.6) does. The reference implementation has a `c` variable that is maintained in the state as a pointer to the buffer. Similarly, the `noble-hashes` implementation of BLAKE2B uses `pos`, which is the same as `buflen`. I’m in favor of following the reference implementation.

Reply:
Ok, I checked another implementation and it does the same. Lgtm